Detect home dir for Windows

elfranne opened the pull request:
Fix for : #559
v5.2 introduced a Linux only method of getting home dir (see #553). 
This PR checks if running on Windows and use the command from 5.1 or else use 5.2+ command.

Tested on Microsoft Windows Server 2019 (64-bit). Maybe @krische or @aaronhilton0 can help testing ?

chelnak commented:
@elfranne Thanks for this!

We don't actually officially support Windows for this module ([see here](https://github.com/puppetlabs/puppetlabs-vcsrepo/blob/main/metadata.json#L13)).

However, thats not to say we can't/shoudn't.

Would you be happy to add support to the manifest and we can see what acceptance tests need looking at.

chelnak commented:
@elfranne Thanks again! I've just kicked the tests off!

Let's see what happens 👀

elfranne commented:
> I have fixed Hiera to use single quotes. There is a issue with Windows: `The password does not meet the password policy requirements. Check the minimum password length, password complexity and password history requirements.`

Not sure how to fix that ... Is it something specific to how Github workflow sets up Windows ?

chelnak commented:
So we aren't clear either at the moment. We will let you know as soon as we've got a fix.

GSPatton commented on metadata.json:
```diff
@@ -11,6 +11,15 @@
   ],
   "operatingsystem_support": [
+    {
+    "operatingsystem": "Windows",
```
this block should be indented

AndreasPfaffeneder commented:
> So we aren't clear either at the moment. We will let you know as soon as we've got a fix.

Would it be sufficient to provide a fixed password in the user-resource in the tests?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/puppetlabs/puppetlabs-vcsrepo?pullRequest=577) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/puppetlabs/puppetlabs-vcsrepo?pullRequest=577) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/puppetlabs/puppetlabs-vcsrepo?pullRequest=577) it.</sub>

LukasAud commented:
This Pull Request needs a rebase and conflict resolution before it can be reviewed again.

elfranne commented:
Fix merge
Fix Lint
Added PR from @AndreasPfaffeneder  (sorry i missed that ...)
Copy `provision.yaml` from stdlib (and added win 2022)

Can someone start the CI workflow to see if that helps ? maybe @LukasAud  ?

elfranne commented:
Anyone got an idea how to make the checks working on Windows ?

elfranne commented:
@ic248 , let's see if this helps ...

Joris29 commented:
Any update on this?

elfranne commented:
> Appears the password worked. Does this just require adding the group to resolve?

let 's try ! Can someone trigger the test ?

smortex commented on provision.yaml:
```diff
@@ -63,4 +63,4 @@ release_checks_7:
   - debian-10-x86_64
   - ubuntu-1804-x86_64
-  - ubuntu-2004-x86_64
+  - ubuntu-2004-x86_64
```
Please revert and check your editor config to avoid this in the future.

elfranne replied:
fixed.

smortex commented on lib/puppet/provider/vcsrepo/git.rb:
```diff
@@ -731,16 +731,29 @@ def git_ssh_with_identity_ssh_file(*args)
-      exec_args[:custom_environment] = { 'HOME' => Etc.getpwnam(@resource.value(:user)).dir }
-      exec_args[:uid] = @resource.value(:user)
+    if Etc.uname[:sysname] == 'Windows_NT'
+      exec_args = {
+        failonfail: true,
+        combine: true,
+        custom_environment: { 'HOME' => Dir.home }
+      }
+      if @resource.value(:user) && @resource.value(:user) != Facter['id'].value
+        exec_args[:custom_environment] = { 'HOME' => Dir.home(@resource.value(:user)) }
+        exec_args[:uid] = @resource.value(:user)
+      end
```
Dir.home seems portable, so using it instead of Etc.getpwnam should do the trick for all platform without the need for a code path different for Windows.

smortex replied:
Ah, only now I see that #553 was about fixing this now working in all circumstances.

Testing with Ruby 3.2, I can't reproduce the described issue:

```sh-session
romain@zappy /home/romain % env -i /usr/local/bin/irb
irb(main):001> ENV
=> {}
irb(main):002> Dir.home
=> "/home/romain"
irb(main):003> 
```

elfranne replied:
So should #553 be reverted ?

smortex replied:
Did more testing running irb with a clean environment:

| Ruby version | `Dir.home` | Comment |
|----|---|----|
| 3.2.3 | "/home/romain" | Puppet 8.5.1 Ruby
| 3.1.2 | "/home/romain" | Debian 12 Ruby
| 2.7.6 | "/home/romain" | Puppet 7.17.0 Ruby
| 2.3.1 | ArgumentError: couldn't find HOME environment -- expanding `~' | Ubuntu 16.04 Ruby |
| 2.0.0 | ArgumentError: couldn't find HOME environment -- expanding `~' | CentOS 7 Ruby


So the version of Ruby shipped with both Puppet 7 and Puppet 8 does not exhibit this problem anymore.  I think the best is just to revert this commit @elfranne .

elfranne replied:
I ll clean it up.

smortex replied:
> I ll clean it up.

Hehe :smile: 

I think the simler for you is basically:

```sh-session
$ git reset --hard origin/main
$ git revert 28c8a65f7cc09ccb647b3c93f80f805b72cd8a96
$ git push -f
```

smortex reviewed (commented):
Some minor comments, but let's see what it does

smortex commented:
@elfranne, I squashed the ~20 commits into a single one and rebased it on top of the main branch to have room to dig into the CI failures.  Unfortunately, the test suite strongly assume a Linux operating system, and I am wondering if attempting to run the test suite on windows actually make sense :confused:?

Joris29 commented:
@elfranne Any updates on this?

Joris29 commented:
@elfranne What's the current status for this?

elfranne commented:
I ll have some time in the upcoming weeks to take a look on this again.

rismoney commented:
For the record, I have cherry picked this and it works on cloning git repos on windows nodes.  I haven't done extensive testing, but core functionality works as expected.
This fixes an error similar to this

Error: Could not set 'present' on ensure: undefined method `dir' for nil:NilClass (file: /etc/puppetlabs/code/environments/mything/modules/server/manifests/mything.pp, line: 12)

elfranne commented:
I ll open a new PR once I get test working.
